src/main/java/gregtech/api/recipes/RecipeMap.java

```diff
@@ -327,95 +296,346 @@ public Recipe findRecipe(long voltage, List<ItemStack> inputs, List<FluidStack>
+        }
+        if (fluids.size() > 0) {
+            List<FluidStack> stack = new ObjectArrayList<>(fluids.size());
+            for (FluidStack f : fluids) {
+                if (f == null || f.amount == 0) {
+                    continue;
+                }
+                stack.add(f);
+            }
+            if (stack.size() > 0) {
+                buildFromFluidStacks(list, stack);
+            }
+        }
+        if (list.size() == 0) {
             return null;
         }
-        return findByInputsAndFluids(voltage, inputs, fluidInputs,exactVoltage);
+        return recurseIngredientTreeFindRecipe(list, lookup, canHandle);
     }
 
-    @Nullable
-    private Recipe findByInputsAndFluids(long voltage, List<ItemStack> inputs, List<FluidStack> fluidInputs, boolean exactVoltage) {
-        HashSet<Recipe> iteratedRecipes = new HashSet<>();
-        HashSet<ItemStackKey> searchedItems = new HashSet<>();
-        HashSet<FluidKey> searchedFluids = new HashSet<>();
-        HashMap<Integer, LinkedList<Recipe>> priorityRecipeMap = new HashMap<>();
-        HashMap<Recipe, Integer> promotedTimes = new HashMap<>();
-
-        for (ItemStack stack : inputs) {
-            if (!stack.isEmpty()) {
-                ItemStackKey itemStackKey = KeySharedStack.getRegisteredStack(stack);
-                if (!searchedItems.contains(itemStackKey) && recipeItemMap.containsKey(itemStackKey)) {
-                    searchedItems.add(itemStackKey);
-                    for (Recipe tmpRecipe : recipeItemMap.get(itemStackKey)) {
-                        if (!exactVoltage && voltage < tmpRecipe.getEUt()) {
-                            continue;
-                        } else if (exactVoltage && voltage != tmpRecipe.getEUt()) {
-                            continue;
-                        }
-                        calculateRecipePriority(tmpRecipe, promotedTimes, priorityRecipeMap);
-                    }
+    /**
+     * Builds a list of unique ItemStacks from the given Collection of ItemStacks.
+     * Used to reduce the number inputs, if for example there is more than one of the same input,
+     * pack them into one.
+     * This uses a strict comparison, so it will not pack the same item with different NBT tags,
+     * to allow the presence of, for example, more than one configured circuit in the input.
+     * @param input The Collection of GTRecipeInputs.
+     * @return an array of unique itemstacks.
+     */
+
+    public static ItemStack[] uniqueItems(Collection<ItemStack> input) {
+        List<ItemStack> list = new ObjectArrayList<>(input.size());
+        loop:
+        for (ItemStack item : input) {
+            if (item.isEmpty()) {
+                continue;
+            }
+            for (ItemStack obj : list) {
+                if (item.isItemEqual(obj) && ItemStack.areItemStackTagsEqual(item, obj)) {
+                    continue loop;
                 }
             }
+            list.add(item);
         }
```

Comment:
You can de-label this segment with the following:

```suggestion
        for (ItemStack item : input) {
            if (item.isEmpty()) {
                continue;
            }
            boolean isEqual = false;
            for (ItemStack obj: list) {
                if (item.isItemEqual(obj) && ItemStack.areItemStackTagsEqual(item, obj)) {
                    isEqual = true;
                    break;
                }
            }
            if (isEqual) continue;
            list.add(item);
        }
```